File: python/monty_hall.py
#!/usr/bin/env python

#Monty Hall Problem
#
# switch = True : change doors
# switch = False: don't change doors

import random
import time
starttime = time.time()

def choose_door_to_open(one, two):
    #If either of the doors is the one with the prize
    if one or two:
        if one:
            return 2
        else:
            return 1
    #If neither of the doors has a prize, return a random number
    else:
        return random.randint(1,2)


wins = 0
losses = 0
rounds = 1000000
switch = True

for i in range(1,rounds+1):
    #Set up doors

    door_one = False
    door_two = False
    door_three = False

    #Choose door that is to have

    door_with_prize = random.randint(1,3)

    if door_with_prize == 1:
        door_one = True
    if door_with_prize == 2:
        door_two = True
    if door_with_prize == 3:
        door_three = True

    #print "Doors:"
    #print "One: "+str(door_one)
    #print "Two: "+str(door_two)
    #print "Three: "+str(door_three)

    #Have the program randomly choose a start door
    chosen_door = random.randint(1,3)
    #print "Chosen door: "+str(chosen_door)

    #Have the program choose which door to open.
    if chosen_door == 1:
        opens = choose_door_to_open(door_two, door_three)
        if opens == 1:
            opens = 2
        else:
            opens = 3
    elif chosen_door == 2:
        opens = choose_door_to_open(door_one, door_three)
        if opens == 1:
            opens = 1
        else:
            opens = 3
    else:
        opens = choose_door_to_open(door_one, door_two)
        if opens == 1:
            opens = 1
        else:
            opens = 2

    #print "Door Opened: "+str(opens)

    #Have the program choose to switch or stay and report result.
    if switch:
        if opens == 1 and chosen_door == 2:
            chosen_door = 3
        elif opens == 1 and chosen_door == 3:
            chosen_door = 2
        elif opens == 2 and chosen_door == 1:
            chosen_door = 3
        elif opens == 2 and chosen_door == 3:
            chosen_door = 1
        elif opens == 3 and chosen_door == 1:
            chosen_door = 2
        elif opens == 3 and chosen_door == 2:
            chosen_door = 1
        #print "Change door to: "+str(chosen_door)
    #else:
        #print "Chosen door is same."

    if chosen_door == door_with_prize:
        #print "Win"
        wins = wins + 1
    else:
        #print "Lose"
        losses = losses + 1

    #print "===End of round: "+str(i)+" ==="

print "===RESULTS==="
print "Wins: "+str(wins)
print str((wins / (rounds * 1.0))*100)+"%"
print "Losses: "+str(losses)
print str((losses / (rounds * 1.0))*100)+"%"
print
#print time.time() - starttime

#raw_input("Press enter to exit...")

File: c/monty_hall.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define true 1
#define false 0
#define SWITCH_DOORS 1
#define ROUNDS 1000000000

//#define PRINTOUT = 1 //Define for printout of text

/* Seed the random with a random value the clock doesn't work because
 * it only gets updated once per second, and rand is called multiple hundreds
 * of times per second.
 */
static int get_rand()
{
    srand(rand());
    return rand();
}

static int choose_door_to_open(int one, int two)
{
    if( one || two )  //If either door actually has the prize
    {
        return (one) ? 2 : 1; //Open the other one
    }

    //If neither door has the prize, just return a random door.
    return get_rand() % 2 + 1;
}

int choose_door_to_switch(int opens, int chosen_door)
{
    if(opens == 1)
        return (chosen_door == 2) ? 3 : 2;

    if(opens == 2)
        return (chosen_door == 1) ? 3 : 1;

    //Opens must be 3, no need for that extra compairison.
    return (chosen_door == 1) ? 2 : 1;
}

int main(int argc, char *argv[])
{
    unsigned int wins = 0;
    unsigned int losses = 0;
    unsigned int i = 1;
    int door_one, door_two, door_three;

    /*Seed the random generator with the time*/
    srand(time(NULL));



    for (i = 1 ; i < ROUNDS + 1; i++)
    {
        //Set up doors
        door_one = false;
        door_two = false;
        door_three = false;

        int door_with_prize = get_rand() % 3 + 1;

        if (door_with_prize == 1)
        {
            door_one = true;
        }
        else
        {
            if (door_with_prize == 2)
                door_two = true;
            else
                door_three = true;
        }

        #ifdef PRINTOUT
        printf("Doors\n");
        printf("One: %i \n", door_one);
        printf("Two: %i\n", door_two);
        printf("Three: %i\n", door_three);
        #endif

        //Have the program randomly choose a start door
        unsigned int chosen_door = get_rand() % 3 + 1;

        #ifdef PRINTOUT
        printf("Chosen Door: %i\n", chosen_door);
        #endif

        //Have the program randomly choose which door to open.
        unsigned int opens;
        if( chosen_door == 1 )
        {
            opens = choose_door_to_open(door_two, door_three);
            opens = (opens == 1) ? 2 : 3;
        }
        else
        {
            if( chosen_door == 2 )
            {
                opens = choose_door_to_open(door_one, door_three);
                opens = (opens == 1) ? 1 : 3;
            }
            else  //Chosen door must be 3
            {
                opens = choose_door_to_open(door_one, door_two);
                opens = (opens == 1) ? 1 : 2;
            }
        }

        #ifdef PRINTOUT
        printf("Door Opened: %i\n", opens);
        #endif

        //Have the program switch doors
        if (SWITCH_DOORS == true)
        {
            chosen_door = choose_door_to_switch(opens, chosen_door);
            #ifdef PRINTOUT
            printf("Change door to: %i \n", chosen_door);
            #endif
        }

        //Increment the proper number so we can do stats at the end.
        if (chosen_door == door_with_prize)
        {
            #ifdef PRINTOUT
            printf( "Win\n" );
            #endif
            wins++;
        }
        else
        {
            #ifdef PRINTOUT
            printf ("Lose\n");
            #endif
            losses++;
        }

        #ifdef PRINTOUT
        printf("===End of round: %i ===\n", i);
        #endif
    }

    printf( "===RESULTS===\n" );
    printf( "Wins: %i\n", wins);
    printf( "%f %% \n",(wins / (ROUNDS * 1.0))*100);
    printf( "Losses: %i\n", losses);
    printf( "%f %% \n",(losses / (ROUNDS * 1.0))*100);

    return 0;
}
